From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Bastiaan van der Plaat <bastiaan.v.d.plaat@gmail.com>
Date: Tue, 6 Oct 2026 07:09:51 +0200
Subject: [PATCH] Fix undefined behavior that modern compilers miscompile

Mod_LoadTexinfo and R_EntityParticles fill two dimensional arrays by
indexing past the end of their first row. GCC assumes that never happens
and cuts the loops short, which would leave the texture vectors and
particle velocities partly unset.

SV_RecursiveHullCheck is called by chase.c without a prototype, so its
float arguments would be passed as doubles.

COM_Init read two bytes of a local array as a short to find the byte
order, which clang can place at an odd address, where the ARM9 can't load
a short.
---
 WinQuake/common.c | 8 +++++---
 WinQuake/model.c  | 5 ++++-
 WinQuake/r_part.c | 7 ++++---
 WinQuake/world.h  | 2 ++
 4 files changed, 15 insertions(+), 7 deletions(-)

diff --git a/WinQuake/common.c b/WinQuake/common.c
index e07e6ce..a958d95 100644
--- a/WinQuake/common.c
+++ b/WinQuake/common.c
@@ -1124,10 +1124,12 @@ COM_Init
 */
 void COM_Init (char *basedir)
 {
-	byte    swaptest[2] = {1,0};
+	short   swaptest = 1;
 
-// set the byte swapping variables in a portable manner 
-	if ( *(short *)swaptest == 1)
+// set the byte swapping variables in a portable manner, the first byte of a
+// short has its low bits on little endian CPUs (the bytes as a short could
+// be misaligned)
+	if ( *(byte *)&swaptest == 1)
 	{
 		bigendien = false;
 		BigShort = ShortSwap;
diff --git a/WinQuake/model.c b/WinQuake/model.c
index e3bd9fb..940f7fe 100644
--- a/WinQuake/model.c
+++ b/WinQuake/model.c
@@ -662,8 +662,11 @@ void Mod_LoadTexinfo (lump_t *l)
 
 	for ( i=0 ; i<count ; i++, in++, out++)
 	{
-		for (j=0 ; j<8 ; j++)
+		for (j=0 ; j<4 ; j++)
+		{
 			out->vecs[0][j] = LittleFloat (in->vecs[0][j]);
+			out->vecs[1][j] = LittleFloat (in->vecs[1][j]);
+		}
 		len1 = Length (out->vecs[0]);
 		len2 = Length (out->vecs[1]);
 		len1 = (len1 + len2)/2;
diff --git a/WinQuake/r_part.c b/WinQuake/r_part.c
index 3db22ab..72af953 100644
--- a/WinQuake/r_part.c
+++ b/WinQuake/r_part.c
@@ -124,7 +124,7 @@ float	timescale = 0.01;
 void R_EntityParticles (entity_t *ent)
 {
 	int			count;
-	int			i;
+	int			i, j;
 	particle_t	*p;
 	float		angle;
 	float		sr, sp, sy, cr, cp, cy;
@@ -136,8 +136,9 @@ void R_EntityParticles (entity_t *ent)
 
 if (!avelocities[0][0])
 {
-for (i=0 ; i<NUMVERTEXNORMALS*3 ; i++)
-avelocities[0][i] = (rand()&255) * 0.01;
+for (i=0 ; i<NUMVERTEXNORMALS ; i++)
+for (j=0 ; j<3 ; j++)
+avelocities[i][j] = (rand()&255) * 0.01;
 }
 
 
diff --git a/WinQuake/world.h b/WinQuake/world.h
index cddb94f..f504d7d 100644
--- a/WinQuake/world.h
+++ b/WinQuake/world.h
@@ -76,3 +76,5 @@ trace_t SV_Move (vec3_t start, vec3_t mins, vec3_t maxs, vec3_t end, int type, e
 // shouldn't be considered solid objects
 
 // passedict is explicitly excluded from clipping checks (normally NULL)
+
+qboolean SV_RecursiveHullCheck (hull_t *hull, int num, float p1f, float p2f, vec3_t p1, vec3_t p2, trace_t *trace);
